### PR TITLE
feat(mappers): combine search response mappers to flatten hierarchical filters

### PR DESCRIPTION
**Pull request by mnavarroespinosa**
[EX-6205](https://searchbroker.atlassian.net/browse/EX-6205)

## Motivation and context

- [ ] Add a response mapper to flatten the filters of the hierarchical facets as expected by `x-components`.

## Type of change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that causes existing functionality to not work as expected)
- [ ] Change requires a documentation update

### CODE REVIEW

**Review by tiborux — APPROVED** [review]

File: packages/x-adapter-platform/src/schemas/response/models/facets/types.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+/**
+ * Facet configuration containing the model name and the schema.
+ */
+export type FacetConfig = { modelName: FacetModelName; getSchema: () => Schema };
+
+/**
+ * Dictionary grouping facets configurations.
+ */
+export type FacetsConfig = Record<string, FacetConfig>;
```
Comment (javieri-empathy) [inline packages/x-adapter-platform/src/schemas/response/models/facets/types.ts:12]: I would suggest to set the property that needs to be the key of the record:

```suggestion
export type FacetsConfig = Record<Facet['id'], FacetConfig>;
```

**Review by javieri-empathy — CHANGES_REQUESTED** [review]

File: packages/x-adapter-platform/src/mappers/response/search-response.mapper.ts [hunk 2]
```diff
@@ -1,8 +1,76 @@
-import { schemaMapperFactory } from '@empathyco/x-adapter-next';
-import { SearchResponse } from '@empathyco/x-types';
+import { schemaMapperFactory, combineMappers, MapperContext } from '@empathyco/x-adapter-next';
+import { SearchResponse, HierarchicalFilter, HierarchicalFacet } from '@empathyco/x-types';
 import { PlatformSearchResponse } from '../../types/responses/search-response.model';
 import { searchResponseMutableSchema } from '../../schemas/response/search-response.schema';
+import { AdapterHierarchicalFacet, AdapterHierarchicalFilter } from '../../types';
```
[the remark this answers is not in the document]
Comment (mnavarroespinosa) [reply packages/x-adapter-platform/src/mappers/response/search-response.mapper.ts]: Good catch, changed!

File: packages/x-adapter-platform/src/schemas/response/models/facet.schema.ts [hunk 3]
```diff
@@ -11,75 +10,63 @@ import {
   PlatformFacet,
   PlatformHierarchicalFilter
 } from '../../../types/responses/models/facet.model';
-import { numberFilterMutableSchema } from './filters/number-filter.schema';
-import { simpleMutableFilterSchema } from './filters/simple-filter.schema';
+import { getFacetConfig, getFacetId } from './facets/utils';
+import { FacetsConfig } from './facets';
+import { numberFilterMutableSchema, simpleMutableFilterSchema } from './filters';
```
[the remark this answers is not in the document]
Comment (mnavarroespinosa) [reply packages/x-adapter-platform/src/schemas/response/models/facet.schema.ts]: Fixed!

File: packages/x-adapter-platform/src/schemas/response/models/facet.schema.ts [hunk 4]
```diff
@@ -11,75 +10,63 @@ import {
...
 
 export const hierarchicalFilterSchema: Schema<PlatformHierarchicalFilter, HierarchicalFilter> = {
-  facetId: 'filter',
+  facetId: (_, $context) => getFacetId($context?.facetId as string),
   label: 'value',
   id: 'filter',
   totalResults: 'count',
-  parentId: (_, $context) => $context?.parentId as string,
+  // eslint-disable-next-line @typescript-eslint/no-extra-parens
+  parentId: (_, $context) => ($context?.isChild ? ($context?.parentId as string) : null),
```
[the remark this answers is not in the document]
Comment (mnavarroespinosa) [reply packages/x-adapter-platform/src/schemas/response/models/facet.schema.ts]: Sure, that makes sense. I've removed the `isChild` property, much simpler now.

File: packages/x-adapter-platform/src/schemas/response/models/facet.schema.ts [hunk 5]
```diff
@@ -11,75 +10,63 @@ import {
...
+  modelName: source => {
+    return getFacetConfig(source.facet).modelName as any;
+  },
   filters: {
     $path: 'values',
-    $subSchema: ({ facet }) => getFilterSchemaFromFacetType(facet),
+    $subSchema: ({ facet }) => getFacetConfig(facet).getSchema() as any,
     $context: {
-      parentId: 'facet'
+      parentId: 'facet',
```
[the remark this answers is not in the document]
Comment (mnavarroespinosa) [reply packages/x-adapter-platform/src/schemas/response/models/facet.schema.ts]: I (mistakenly) thought it was necessary to get the parent facet id from the `hierarchicalFilterSchema`. I've removed it, now the `parentId` is added to the context from the `hierarchicalFilterSchema`.

File: packages/x-adapter-platform/src/schemas/response/models/facets/types.ts (hunk 1 above)
Comment (mnavarroespinosa) [reply packages/x-adapter-platform/src/schemas/response/models/facets/types.ts]: Sure, thanks! I've changed it.

File: packages/x-adapter-platform/src/schemas/response/models/facets/types.ts [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import { FacetModelName } from '@empathyco/x-types';
+import { Schema } from '@empathyco/x-adapter-next';
+
+/**
+ * Facet configuration containing the model name and the schema.
+ */
+export type FacetConfig = { modelName: FacetModelName; getSchema: () => Schema };
```
[the remark this answers is not in the document]
Comment (mnavarroespinosa) [reply packages/x-adapter-platform/src/schemas/response/models/facets/types.ts]: There was no particular reason, seems great to me to keep the same code style. I've changed it!

File: packages/x-adapter-platform/src/schemas/response/models/facets/utils.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * Returns the facet's config.
+ *
+ * @param facet - The facet to resolve the configuration.
+ * @returns The facet's config.
+ */
+export function getFacetConfig(facet: string): FacetConfig {
+  const name = getFacetId(facet);
+  return facetsConfig[name] ? facetsConfig[name] : facetsConfig.default;
```
[the remark this answers is not in the document]
Comment (mnavarroespinosa) [reply packages/x-adapter-platform/src/schemas/response/models/facets/utils.ts]: Changed!

File: packages/x-adapter-platform/src/mappers/response/search-response.mapper.ts [hunk 8]
```diff
@@ -1,8 +1,76 @@
...
+ * @param initialValues - The initial values for the filter.
+ * @param filters - The filters array to fill with the facet filters.
+ * @returns The filter id.
+ */
+function mapHierarchicalFilter(
+  rawFilter: AdapterHierarchicalFilter,
+  initialValues: Readonly<Partial<HierarchicalFilter>>,
+  filters: HierarchicalFilter[]
+): HierarchicalFilter['id'] {
+  const filter = { ...rawFilter, ...initialValues } as HierarchicalFilter;
```
[the remark this answers is not in the document]
Comment (mnavarroespinosa) [reply packages/x-adapter-platform/src/mappers/response/search-response.mapper.ts]: Sure! You're totally right, it's not a good idea to do this casting. Changed!

**Pull request closed without merging**